Closing of websocket by remote server does not trigger the script close handler

When the remote server closes the websocket [like this](https://github.com/loadimpact/k6/blob/d6f9a55ad1ebb131b1d3e65a05ce32bd68c8ab8d/lib/testutils/httpmultibin.go#L92), the `close` handler in the script is not called and the script iteration never finishes executing, it becomes stuck in the websocket handling code. 

The `error` handler is called with an error like `GoError: An unexpected error occured: websocket: close 1006 (abnormal closure): unexpected EOF....`, which is probably also suboptimal.

On related note I just got a panic in the ws test: 
```
ok      github.com/loadimpact/k6/js/modules/k6/metrics  4.385s
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 4: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 4: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 7: Panic at 4: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 7: Panic at 4: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: Panic at 7: Panic at 4: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x83cb53]

goroutine 6 [running]:
testing.tRunner.func1(0xc000132900)
        /home/mstoykov/.gvm/gos/go1.11.1/src/testing/testing.go:792 +0x387
panic(0x99bbc0, 0xc000a906f0)
        /home/mstoykov/.gvm/gos/go1.11.1/src/runtime/panic.go:513 +0x1b9
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunProgram.func1(0xc003085e10)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:880 +0x9b
panic(0x99bbc0, 0xc000a906f0)
        /home/mstoykov/.gvm/gos/go1.11.1/src/runtime/panic.go:513 +0x1b9
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try.func1(0xc0001daa90, 0x0, 0xc003085d10, 0x0, 0x0, 0x0, 0xc003085d68)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:364 +0x459
panic(0x99bbc0, 0xc000a906e0)
        /home/mstoykov/.gvm/gos/go1.11.1/src/runtime/panic.go:513 +0x1b9
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1.1(0xc003084f80)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1407 +0x9b
panic(0x99bbc0, 0xc000a906e0)
        /home/mstoykov/.gvm/gos/go1.11.1/src/runtime/panic.go:513 +0x1b9
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try.func1(0xc0001daa90, 0x1, 0xc003084e88, 0x4, 0x0, 0x1, 0xc003084ee0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:364 +0x459
panic(0x9a9640, 0x1139e60)
        /home/mstoykov/.gvm/gos/go1.11.1/src/runtime/panic.go:513 +0x1b9
github.com/loadimpact/k6/vendor/github.com/gorilla/websocket.(*Conn).WriteControl(0x0, 0x8, 0xc00089e158, 0x2, 0x2, 0xbf16569b8b8a907a, 0x463dda2b8, 0x114c640, 0x0, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/gorilla/websocket/conn.go:412 +0x73
github.com/loadimpact/k6/js/modules/k6/ws.(*Socket).closeConnection.func1()
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/modules/k6/ws/ws.go:434 +0x16d
sync.(*Once).Do(0xc0004e2170, 0xc003084550)
        /home/mstoykov/.gvm/gos/go1.11.1/src/sync/once.go:44 +0xb3
github.com/loadimpact/k6/js/modules/k6/ws.(*Socket).closeConnection(0xc0004e2140, 0x3e9, 0x465eb6, 0x973b40)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/modules/k6/ws/ws.go:431 +0x76
github.com/loadimpact/k6/js/modules/k6/ws.(*Socket).Close(0xc0004e2140, 0x116afc0, 0x0, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/modules/k6/ws/ws.go:424 +0x3f
reflect.callMethod(0xc0021720c0, 0xc0030846b0, 0xc003084698)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/value.go:689 +0x180
reflect.methodValueCall(0x116afc0, 0x0, 0x0, 0xc0014e6580, 0xc003084998, 0x4b4db9, 0xc000fa24e0, 0xc0021720c0, 0xc0014e65c0, 0x1800000018, ...)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/asm_amd64.s:35 +0x42
reflect.Value.call(0x971880, 0xc0021720c0, 0x13, 0xa52844, 0x4, 0x116afc0, 0x1, 0x0, 0xa50020, 0x1, ...)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/value.go:447 +0x449
reflect.Value.Call(0x971880, 0xc0021720c0, 0x13, 0x116afc0, 0x0, 0x0, 0x203000, 0x203000, 0xc0014e6540)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/value.go:308 +0xa4
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).wrapReflectFunc.func1(0xbb8420, 0xc0014e61c0, 0xc0004e8090, 0x0, 0x5, 0xc0004e80e0, 0x2)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1153 +0x7c3
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm)._nativeCall(0xc0001daa90, 0xc0004e4210, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1826 +0x284
github.com/loadimpact/k6/vendor/github.com/dop251/goja.call.exec(0x0, 0xc0001daa90)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1810 +0x4e3
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).run(0xc0001daa90)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:288 +0x5d
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*funcObject).Call(0xc000bd6480, 0xbb8b00, 0x116ad20, 0xc000a90500, 0x1, 0x1, 0xc003084dd8, 0x40beb3)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/func.go:130 +0x2be
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*funcObject).Call-fm(0xbb8b00, 0x116ad20, 0xc000a90500, 0x1, 0x1, 0xa07200, 0xc0004e2140)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:120 +0x48
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1.2()
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1412 +0x96
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try(0xc0001daa90, 0xc003084ef8, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:370 +0x10c
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1(0xbb8b00, 0x116ad20, 0xc000a90500, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1411 +0x10d
github.com/loadimpact/k6/js/modules/k6/ws.(*WS).Connect(0x116afc0, 0xbb3da0, 0xc000209e60, 0xc00107e6a4, 0x1b, 0xc000a8ea30, 0x1, 0x1, 0x0, 0x0, ...)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/modules/k6/ws/ws.go:196 +0x70f
reflect.Value.call(0x99bc20, 0x116afc0, 0x213, 0xa5489c, 0x9, 0xc000bdc410, 0x3, 0x3, 0xa0eac0, 0xc000a8ea40, ...)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/value.go:447 +0x449
reflect.Value.CallSlice(0x99bc20, 0x116afc0, 0x213, 0xc000bdc410, 0x3, 0x3, 0x194, 0x8, 0xc0030859c8)
        /home/mstoykov/.gvm/gos/go1.11.1/src/reflect/value.go:321 +0xa4
github.com/loadimpact/k6/js/common.Bind.func1(0xbb8420, 0xc00020afc0, 0xc000c389e0, 0x2, 0x4, 0xc0021887c0, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/common/bridge.go:237 +0xa23
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm)._nativeCall(0xc0001daa90, 0xc000bd2f20, 0x2)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1826 +0x284
github.com/loadimpact/k6/vendor/github.com/dop251/goja.call.exec(0x2, 0xc0001daa90)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1810 +0x4e3
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).run(0xc0001daa90)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:288 +0x5d
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).run-fm()
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:375 +0x2a
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try(0xc0001daa90, 0xc003085d70, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:370 +0x10c
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).runTry(0xc0001daa90, 0xb376d0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:375 +0x4e
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunProgram(0xc000001b00, 0xc000c388a0, 0x0, 0x0, 0x0, 0x0)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:891 +0x1b7
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunScript(0xc000001b00, 0x0, 0x0, 0xc00107e4e0, 0xc4, 0x0, 0x8000106, 0x0, 0xffffffffffffffff)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:870 +0x9b
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).RunString(0xc000001b00, 0xc00107e4e0, 0xc4, 0xa5559a, 0xa, 0xc00107e4e0, 0xc4)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:859 +0x47
github.com/loadimpact/k6/js/common.RunString(0xc000001b00, 0xa6f074, 0xc1, 0xc000052798, 0x47cb36, 0x5c78e2db, 0x14b40b0b)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/common/util.go:39 +0xf3
github.com/loadimpact/k6/js/modules/k6/ws.TestSession.func1(0xc000132900)
        /home/mstoykov/.gvm/pkgsets/go1.11.1/global/src/github.com/loadimpact/k6/js/modules/k6/ws/ws_test.go:122 +0x3f
testing.tRunner(0xc000132900, 0xc000201240)
        /home/mstoykov/.gvm/gos/go1.11.1/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
        /home/mstoykov/.gvm/gos/go1.11.1/src/testing/testing.go:878 +0x353
FAIL    github.com/loadimpact/k6/js/modules/k6/ws       8.876s
```
This has never happened before and I do a lot of test runs :) I can't at the moment look into it but it might be relevant

--- a/core/engine_test.go
+++ b/core/engine_test.go
@@ -502,6 +502,7 @@ func getMetricCount(collector *dummy.Collector, name string) (result uint) {
 
 const expectedHeaderMaxLength = 500
 
+// FIXME: This test is too brittle, consider simplifying.
 func TestSentReceivedMetrics(t *testing.T) {
 	t.Parallel()
 	tb := httpmultibin.NewHTTPMultiBin(t)
@@ -532,10 +533,14 @@ func TestSentReceivedMetrics(t *testing.T) {
 					file: http.file(data, "test.txt")
 				});
 			}`), 1, 1000, 100},
+		// NOTE(imiric): This needs to keep testing against /ws-echo-invalid because
+		// this test is highly sensitive to metric data, and slightly differing
+		// WS server implementations might introduce flakiness.
+		// See https://github.com/loadimpact/k6/pull/1149
 		{tr(`import ws from "k6/ws";
 			let data = "0123456789".repeat(100);
 			export default function() {
-				ws.connect("WSBIN_URL/ws-echo", null, function (socket) {
+				ws.connect("WSBIN_URL/ws-echo-invalid", null, function (socket) {
 					socket.on('open', function open() {
 						socket.send(data);
 					});

--- a/js/modules/k6/ws/ws.go
+++ b/js/modules/k6/ws/ws.go
@@ -223,6 +223,13 @@ func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) (*WSHTTP
 	// The connection is now open, emit the event
 	socket.handleEvent("open")
 
+	// Make the default close handler a noop to avoid duplicate closes,
+	// since we use custom closing logic to call user's event
+	// handlers and for cleanup. See closeConnection.
+	// closeConnection is not set directly as a handler here to
+	// avoid race conditions when calling the Goja runtime.
+	conn.SetCloseHandler(func(code int, text string) error { return nil })
+
 	// Pass ping/pong events through the main control loop
 	pingChan := make(chan string)
 	pongChan := make(chan string)
@@ -262,9 +269,8 @@ func (*WS) Connect(ctx context.Context, url string, args ...goja.Value) (*WSHTTP
 		case readErr := <-readErrChan:
 			socket.handleEvent("error", rt.ToValue(readErr))
 
-		case readClose := <-readCloseChan:
-			// handle server close
-			socket.handleEvent("close", rt.ToValue(readClose))
+		case code := <-readCloseChan:
+			_ = socket.closeConnection(code)
 
 		case scheduledFn := <-socket.scheduled:
 			if _, err := scheduledFn(goja.Undefined()); err != nil {
@@ -425,7 +431,8 @@ func (s *Socket) Close(args ...goja.Value) {
 	_ = s.closeConnection(code)
 }
 
-// Attempts to close the websocket gracefully
+// closeConnection cleanly closes the WebSocket connection.
+// Returns an error if sending the close control frame fails.
 func (s *Socket) closeConnection(code int) error {
 	var err error
 
@@ -437,15 +444,16 @@ func (s *Socket) closeConnection(code int) error {
 			time.Now().Add(writeWait),
 		)
 		if err != nil {
-			// Just call the handler, we'll try to close the connection anyway
+			// Call the user-defined error handler
 			s.handleEvent("error", rt.ToValue(err))
 		}
 
-		// trigger `close` event when the client closes the connection
+		// Call the user-defined close handler
 		s.handleEvent("close", rt.ToValue(code))
+
 		_ = s.conn.Close()
 
-		// Stops the main control loop
+		// Stop the main control loop
 		close(s.done)
 	})
 
@@ -454,21 +462,19 @@ func (s *Socket) closeConnection(code int) error {
 
 // Wraps conn.ReadMessage in a channel
 func readPump(conn *websocket.Conn, readChan chan []byte, errorChan chan error, closeChan chan int) {
-	defer func() { _ = conn.Close() }()
-
 	for {
 		_, message, err := conn.ReadMessage()
 		if err != nil {
-
-			if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
-				closeChan <- err.(*websocket.CloseError).Code
-			} else if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
-				// Emit the error if it is not CloseNormalClosure
-				// and the error is not  originated from closing the socket ourselves with `CloseGoingAway`
+			if websocket.IsUnexpectedCloseError(
+				err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+				// Report an unexpected closure
 				errorChan <- err
 			}
-
-			//CloseGoingAway errors are ignored
+			code := websocket.CloseGoingAway
+			if e, ok := err.(*websocket.CloseError); ok {
+				code = e.Code
+			}
+			closeChan <- code
 			return
 		}
 

--- a/js/modules/k6/ws/ws_test.go
+++ b/js/modules/k6/ws/ws_test.go
@@ -22,10 +22,15 @@ package ws
 import (
 	"context"
 	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
 	"strconv"
 	"testing"
+	"time"
 
 	"github.com/dop251/goja"
+	"github.com/gorilla/websocket"
 	"github.com/loadimpact/k6/js/common"
 	"github.com/loadimpact/k6/lib"
 	"github.com/loadimpact/k6/lib/metrics"
@@ -264,7 +269,7 @@ func TestSession(t *testing.T) {
 	assertSessionMetricsEmitted(t, samplesBuf, "", sr("WSBIN_URL/ws-echo"), 101, "")
 	assertMetricEmitted(t, metrics.WSPing, samplesBuf, sr("WSBIN_URL/ws-echo"))
 
-	t.Run("close", func(t *testing.T) {
+	t.Run("client_close", func(t *testing.T) {
 		_, err := common.RunString(rt, sr(`
 		let closed = false;
 		let res = ws.connect("WSBIN_URL/ws-echo", function(socket){
@@ -280,6 +285,36 @@ func TestSession(t *testing.T) {
 		assert.NoError(t, err)
 	})
 	assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSBIN_URL/ws-echo"), 101, "")
+
+	serverCloseTests := []struct {
+		name     string
+		endpoint string
+	}{
+		{"server_close_ok", "/ws-echo"},
+		// Ensure we correctly handle invalid WS server
+		// implementations that close the connection prematurely
+		// without sending a close control frame first.
+		{"server_close_invalid", "/ws-close-invalid"},
+	}
+
+	for _, tc := range serverCloseTests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := common.RunString(rt, sr(fmt.Sprintf(`
+			let closed = false;
+			let res = ws.connect("WSBIN_URL%s", function(socket){
+				socket.on("open", function() {
+					socket.send("test");
+				})
+				socket.on("close", function() {
+					closed = true;
+				})
+			});
+			if (!closed) { throw new Error ("close event not fired"); }
+			`, tc.endpoint)))
+			assert.NoError(t, err)
+		})
+	}
 }
 
 func TestErrors(t *testing.T) {
@@ -332,7 +367,7 @@ func TestErrors(t *testing.T) {
 
 	t.Run("error_in_setup", func(t *testing.T) {
 		_, err := common.RunString(rt, sr(`
-		let res = ws.connect("WSBIN_URL/ws-echo", function(socket){
+		let res = ws.connect("WSBIN_URL/ws-echo-invalid", function(socket){
 			throw new Error("error in setup");
 		});
 		`))
@@ -342,7 +377,7 @@ func TestErrors(t *testing.T) {
 	t.Run("send_after_close", func(t *testing.T) {
 		_, err := common.RunString(rt, sr(`
 		let hasError = false;
-		let res = ws.connect("WSBIN_URL/ws-echo", function(socket){
+		let res = ws.connect("WSBIN_URL/ws-echo-invalid", function(socket){
 			socket.on("open", function() {
 				socket.close();
 				socket.send("test");
@@ -357,7 +392,7 @@ func TestErrors(t *testing.T) {
 		}
 		`))
 		assert.NoError(t, err)
-		assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSBIN_URL/ws-echo"), 101, "")
+		assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSBIN_URL/ws-echo-invalid"), 101, "")
 	})
 
 	t.Run("error on close", func(t *testing.T) {
@@ -513,4 +548,57 @@ func TestTLSConfig(t *testing.T) {
 		assert.NoError(t, err)
 	})
 	assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSSBIN_URL/ws-close"), 101, "")
+}
+
+func TestReadPump(t *testing.T) {
+	var closeCode int
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		conn, err := (&websocket.Upgrader{}).Upgrade(w, r, w.Header())
+		assert.NoError(t, err)
+		closeMsg := websocket.FormatCloseMessage(closeCode, "")
+		_ = conn.WriteControl(websocket.CloseMessage, closeMsg, time.Now().Add(time.Second))
+	}))
+	defer srv.Close()
+
+	closeCodes := []int{websocket.CloseNormalClosure, websocket.CloseGoingAway, websocket.CloseInternalServerErr}
+
+	numAsserts := 0
+	srvURL := "ws://" + srv.Listener.Addr().String()
+
+	// Ensure readPump returns the response close code sent by the server
+	for _, code := range closeCodes {
+		code := code
+		t.Run(strconv.Itoa(code), func(t *testing.T) {
+			closeCode = code
+			conn, resp, err := websocket.DefaultDialer.Dial(srvURL, nil)
+			assert.NoError(t, err)
+			defer func() {
+				_ = resp.Body.Close()
+				_ = conn.Close()
+			}()
+
+			msgChan := make(chan []byte)
+			errChan := make(chan error)
+			closeChan := make(chan int)
+			go readPump(conn, msgChan, errChan, closeChan)
+
+		readChans:
+			for {
+				select {
+				case responseCode := <-closeChan:
+					assert.Equal(t, code, responseCode)
+					numAsserts++
+					break readChans
+				case <-errChan:
+					continue
+				case <-time.After(time.Second):
+					t.Errorf("Read timed out")
+					break readChans
+				}
+			}
+		})
+	}
+
+	// Ensure all close code asserts passed
+	assert.Equal(t, numAsserts, len(closeCodes))
 }

--- a/lib/testutils/httpmultibin/httpmultibin.go
+++ b/lib/testutils/httpmultibin/httpmultibin.go
@@ -100,32 +100,42 @@ type jsonBody struct {
 	Compression string      `json:"compression"`
 }
 
-func websocketEchoHandler(w http.ResponseWriter, req *http.Request) {
-	conn, err := (&websocket.Upgrader{}).Upgrade(w, req, w.Header())
-	if err != nil {
-		return
-	}
-
-	mt, message, err := conn.ReadMessage()
-	if err != nil {
-		return
-	}
-	err = conn.WriteMessage(mt, message)
-	if err != nil {
-		return
-	}
-	err = conn.Close()
-	if err != nil {
-		return
-	}
-}
-
-func websocketCloserHandler(w http.ResponseWriter, req *http.Request) {
-	conn, err := (&websocket.Upgrader{}).Upgrade(w, req, w.Header())
-	if err != nil {
-		return
-	}
-	_ = conn.Close()
+func getWebsocketHandler(echo bool, closePrematurely bool) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		conn, err := (&websocket.Upgrader{}).Upgrade(w, req, w.Header())
+		if err != nil {
+			return
+		}
+		if echo {
+			messageType, r, e := conn.NextReader()
+			if e != nil {
+				return
+			}
+			var wc io.WriteCloser
+			wc, err = conn.NextWriter(messageType)
+			if err != nil {
+				return
+			}
+			if _, err = io.Copy(wc, r); err != nil {
+				return
+			}
+			if err = wc.Close(); err != nil {
+				return
+			}
+		}
+		// closePrematurely=true mimics an invalid WS server that doesn't
+		// send a close control frame before closing the connection.
+		if !closePrematurely {
+			closeMsg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+			_ = conn.WriteControl(websocket.CloseMessage, closeMsg, time.Now().Add(time.Second))
+			// Wait for response control frame
+			<-time.After(time.Second)
+		}
+		err = conn.Close()
+		if err != nil {
+			return
+		}
+	})
 }
 
 func writeJSON(w io.Writer, v interface{}) error {
@@ -193,8 +203,10 @@ func NewHTTPMultiBin(t testing.TB) *HTTPMultiBin {
 	// Create a http.ServeMux and set the httpbin handler as the default
 	mux := http.NewServeMux()
 	mux.Handle("/brotli", getEncodedHandler(t, httpext.CompressionTypeBr))
-	mux.HandleFunc("/ws-echo", websocketEchoHandler)
-	mux.HandleFunc("/ws-close", websocketCloserHandler)
+	mux.Handle("/ws-echo", getWebsocketHandler(true, false))
+	mux.Handle("/ws-echo-invalid", getWebsocketHandler(true, true))
+	mux.Handle("/ws-close", getWebsocketHandler(false, false))
+	mux.Handle("/ws-close-invalid", getWebsocketHandler(false, true))
 	mux.Handle("/zstd", getEncodedHandler(t, httpext.CompressionTypeZstd))
 	mux.Handle("/zstd-br", getZstdBrHandler(t))
 	mux.Handle("/", httpbin.New().Handler())

